Feature/azure identity

davidatorres opened the pull request:
## Description

This enhancement gives **Mem0** greater reach into organizations with strict security policies that do not allow Azure resources to be accessed via API keys.

The `DefaultAzureCredential` class in the Microsoft Azure Identity library (`azure.identity`) is designed to simplify authentication with Azure services based on the environment the application is running. It does this by **chaining together multiple credential types** and attempting them in a specific order until the first successful authentication.

**Order of Authentication Attempts:**
1. **EnvironmentCredential:**
Looks for Azure client ID, secret, tenant ID, or certificate in environment variables for service principal authentication.
2. **WorkloadIdentityCredential:**
Utilizes Azure Workload Identity (relevant for Kubernetes and Azure workloads).
3. **ManagedIdentityCredential:**
Authenticates as a Managed Identity (for apps/services hosted in Azure with Managed Identity enabled, or a User-assigned Managed identity)
4. **SharedTokenCacheCredential / VisualStudioCredential (Windows only):**
Uses cached credentials from Visual Studio sign-ins (and sometimes VS Code if SSO is enabled).
5. **AzureCliCredential:**
Uses the user session from an Azure CLI login (`az login`).
6. **AzurePowerShellCredential:**
Uses the identity from the Azure PowerShell command (`Connect-AzAccount`).
7. **AzureDeveloperCliCredential:**
Uses the user session from an Azure Developer CLI login (`azd auth login`).

For local development, `DefaultAzureCredential` will commonly succeed with `AzureCliCredential` if you're signed in with `az login`. In deployed environments, it typically uses a managed identity or environment credentials.

***Why Use `DefaultAzureCredential`?***

- **Seamless local-to-cloud transitions:** Write code once and authenticate reliably both locally and in production, with no need to rewrite credential logic for different environments.
- **Supports developer tools:** Recognizes sign-ins from Azure CLI, PowerShell, Developer CLI, Visual Studio, and more.

The LLM and Embedder modules for Azure OpenAI have been updated to utilize the DefaultAzureCreditial() class when the API key is not present in the Memory configuration.  Mem0 authentication for Azure OpenAI retains its default authentication choice of an API key,  this enhancement extends the authentication capability.

The Vector Store module for Azure AI Search has been updated to utilize the DefaultAzureCreditial() class when the API key is not present in the Memory configuration.  Mem0 authentication for Azure AI Search retains its default authentication choice of an API key,  this enhancement extends the authentication capability.

Example notebooks were added to the project to demostrate and test this feature enhancement.

Fixes #3242

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] Documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

A sample python application that wrote three chat messages to memory using Azure OpenAI for embedding (text-embedding-3-large) and chat (gpt-4o), with Azure AI Search for the vector store. The code was tested with the default API key authentication to ensure that the modules retain their default behavior.  The Azure Identity credential chain testing include tests using EnvironmentCredential, ManagedIndentityCredential (user-assigned) and ManagedIdentityCredential (system-assigned), which are typically used in a production environments. Plus, AzureCliCredential and AzureDeveloperCliCredential were used to test locally on a Mac (Apple Silicon) and Windows 11 workstation.

Documentation has been updated to provide direction on how to set up RBAC roles, necessary Environment variables, and how to use developer command-line logins. 

Unit tests were added to test all variations of the Mem0 configuration.

All developer instructions related to pre-commit and stand-alone linting and linting were followed.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [X] closes #3242
- [X] Made sure Checks passed

davidatorres commented:
@blino @mark-watson 

What am I missing to get the PR reviewed?

parshvadaftari commented on pyproject.toml:
```diff
@@ -18,6 +18,7 @@ dependencies = [
     "posthog>=3.5.0",
     "pytz>=2024.1",
     "sqlalchemy>=2.0.31",
+    "azure-identity>=1.23.0",
```
No need to add the dependency over here. This is the default dependencies. Add it to the vector_stores dependency.

davidatorres replied:
Moved dependency to vector_stores

parshvadaftari commented on docs/components/embedders/models/azure_openai.mdx:
```diff
@@ -77,6 +77,92 @@ await memory.add(messages, { userId: "john" });
+7. **Azure Developer CLI Credential:**
+Uses the session from Azure Developer CLI (`azd auth login`).
+
+<Note> If an API is provided, it will be used for authentication over an Azure Identity </Note>
```
Looks good from here and valid as well.

parshvadaftari commented on docs/components/embedders/models/azure_openai.mdx:
````diff
@@ -77,6 +77,92 @@ await memory.add(messages, { userId: "john" });
 ```
 </CodeGroup>
 
+As an alternative to using an API key, the Azure Identity credential chain can be used to authenticate with Azure OpenAI. The list below shows the order of precedence for credential application:
````
This section seems irrelevant to the mem0 docs and is more aligned with how azure handles credential chain.

parshvadaftari commented on docs/components/llms/models/azure_openai.mdx:
````diff
@@ -118,6 +120,92 @@ config = {
 ```
 
+As an alternative to using an API key, the Azure Identity credential chain can be used to authenticate with Azure OpenAI. The list below shows the order of precedence for credential application:
+
````
Same goes here as well

davidatorres replied:
Removed Azure Identity setup details from the documentation.

parshvadaftari reviewed (commented):
Try to refactor the docs as requested.

parshvadaftari commented:
@davidatorres The tests are failing as well.

parshvadaftari commented:
@davidatorres Please incorporate the requested changes. And also see why the tests are failing.

davidatorres commented:
- Updated version to azure-identity>=1.24.0, moved into vector_store section.
- Updated failed pytests, ensured all tests passed for python 3.9, 3.10, 3.11, 3.12
- Revised documentation, removing Azure Identity setup content.

@parshvadaftari - Please let me know there are any additional adjustments needed.

parshvadaftari commented on examples/azure-resources/azure_api_key.ipynb:
```diff
@@ -0,0 +1,230 @@
+{
```
You can remove these notebook files. As we are already adding it to the docs.

davidatorres replied:
Done

parshvadaftari reviewed (commented):
Rest of it looks good to me.

parshvadaftari commented:
@davidatorres Thank you for contributing. Feel free to use it out.
